### PR TITLE
Lazily compute signature type predicates

### LINKED ISSUE #17451: Strange stateful behavior with inferred type predicate type
**TypeScript Version:**  nightly (2.5.0-dev.20170725)

**Code**

```ts
function tryCast<T, U extends T>(value: T, predicate: (t: T) => t is U): U | undefined {
    return predicate(value) ? value : undefined;
}

declare const sn: string | number;

const x: number | undefined = tryCast(sn, (n): n is number => typeof n === "number"); // Always works

const y = tryCast(sn, (n): n is number => typeof n === "number");

const z: number | undefined = y;
```

**Expected behavior:**

No error.

**Actual behavior:**

```
src/a.ts(11,7): error TS2322: Type 'string | number | undefined' is not assignable to type 'number | undefined'.
  Type 'string' is not assignable to type 'number | undefined'.
```

**Depending on the order** I hover over things, hovering over `tryCast` and `y` might show a type of `string | number | undefined`, but it might show a type of `number | undefined`. (But the error at `z` is always present.)

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #17451
**EDIT**: And #19640 and #19642 and #20186.

To summarize the problem:

During the *first* quickInfo, we start from a blank slate.

In  `chooseOverload`:
* First time around, we set `excludedArgument` to `[true]`, meaning that we exclude the sole argument.
* Obviously, we get no type inferences from this. So the inferred type argument is `{}`.
* Then in the *next* round, `excludeArgument` will be `undefined`. This means we enter the body of `checkFunctionExpressionOrObjectLiteralMethod`; particularly where we call `instantiateSignature(contextualSignature, contextualMapper);`.
* In instantiating the contextual signature, we end up calling an innocent-looking function called `cloneTypePredicate`. This uses the contextual mapper to call map the contextual type predicate type to the actual type predicate type.
* By calling the `mapper` (created by `createInferenceContext`), we cause a side-effect of fixing the inference -- so it gets stuck at `{}` forever.

The statefulness in the original issue is because by the *second* quickInfo, we've already checked the function, and its return type is `(n: {}) => n is number`. The explicit `: n is number` annotation overrides the contextual type. So `getTypeOfFuncClassEnumModule` returns a completed full type just an empty new one, and we can infer types correctly. During the first quickInfo we didn't have a type for the function yet, so type inference problems only showed up then.

We had already solved this sort of problem for contextual return types: compute them lazily so that we don't call the contextual mapper until necessary. So the solution is simply to do the same for the type predicate.

### CODE REVIEW

**Comment by andy-ms** [comment]
For posterity, the test I was using to reproduce #17451 was:

```
///<reference path="fourslash.ts"/>

// @noLib: true

////declare function f<T>(predicate: (t: {}) => t is T): T;
////const y = /**/f((n): n is number => true);

verify.quickInfoAt("", "function f<{}>(predicate: (t: {}) => t is {}): {}"); // OK

console.log("First succeeded...");

verify.quickInfoAt("", "function f<{}>(predicate: (t: {}) => t is {}): {}"); // Fails
```

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -6536,6 +6537,17 @@ namespace ts {
             }
         }
 
+        function signatureHasTypePredicate(signature: Signature): boolean {
+            return signature.resolvedTypePredicate !== undefined;
+        }
+
+        function getTypePredicateOfSignature(signature: Signature): TypePredicate {
+            if (signature.resolvedTypePredicate === "pending") {
```
Comment (sandersn) [inline src/compiler/checker.ts]: oh, this is documented in the definition in types.ts

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -5434,14 +5435,14 @@ namespace ts {
         }
 
         function createSignature(declaration: SignatureDeclaration, typeParameters: TypeParameter[], thisParameter: Symbol | undefined, parameters: Symbol[],
-            resolvedReturnType: Type, typePredicate: TypePredicate, minArgumentCount: number, hasRestParameter: boolean, hasLiteralTypes: boolean): Signature {
+            resolvedReturnType: Type | undefined, resolvedTypePredicate: TypePredicate | undefined | "pending", minArgumentCount: number, hasRestParameter: boolean, hasLiteralTypes: boolean): Signature {
```
Comment (rbuckton) [inline src/compiler/checker.ts]: Please be aware that if we call `createSignature` once with a `TypePredicate` and then call `createSignature` again later with `"pending"` that it will result in a V8 compiling a second copy of `createSignature` due to the parameter being polymorphic (string vs. object).

We also end up making `Signature` more polymorphic. This could result in degraded performance so I would recommend you run benchmarks before and after this change to ensure this does not cause a regression.

As an alternative, I would suggest that you define something like a `const unresolvedTypePredicate = <IdentifierTypePredicate>{ kind: TypePredicateKind.Identifier, parameterName: "pending", parameterIndex: 0 };` to act as your `"pending"` sentinel value as it will reduce polymorphism.

Comment (andy-ms) [inline src/compiler/checker.ts]: Could we just use `null` for pending?

Comment (rbuckton) [inline src/compiler/checker.ts]: We try not to use null in the compiler at all, and it would be inconsistent.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -280,6 +280,11 @@ namespace ts {
         const resolvingSignature = createSignature(undefined, undefined, undefined, emptyArray, anyType, /*typePredicate*/ undefined, 0, /*hasRestParameter*/ false, /*hasLiteralTypes*/ false);
         const silentNeverSignature = createSignature(undefined, undefined, undefined, emptyArray, silentNeverType, /*typePredicate*/ undefined, 0, /*hasRestParameter*/ false, /*hasLiteralTypes*/ false);
 
+        const unresolvedTypePredicate: void & { __unresolvedTypePredicate: void } = (() => {
```
Comment (rbuckton) [inline src/compiler/checker.ts]: Why is this type so needlessly complex? Just use `TypePredicate`.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -11403,7 +11423,7 @@ namespace ts {
                     const apparentType = getApparentType(funcType);
                     if (apparentType !== unknownType) {
                         const callSignatures = getSignaturesOfType(apparentType, SignatureKind.Call);
-                        return !!forEach(callSignatures, sig => sig.typePredicate);
+                        return some(callSignatures, sig => signatureHasTypePredicate(sig));
```
Comment (rbuckton) [inline src/compiler/checker.ts]: Can you just use `signatureHasTypePredicate` directly instead?

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -6543,6 +6549,17 @@ namespace ts {
             }
         }
 
+        function signatureHasTypePredicate(signature: Signature): boolean {
+            return signature.resolvedTypePredicate !== undefined;
+        }
+
+        function getTypePredicateOfSignature(signature: Signature): TypePredicate | undefined {
```
Comment (rbuckton) [inline src/compiler/checker.ts]: Another approach we could consider is the one used to resolve Type Parameter constraints. Instead of using `undefined` to indicate no type predicate and an `unresolvedTypePredicate` sentinel to indicate deferred resolution, use `undefined` to indicate deferred resolution and a `noTypePredicate` sentinel. See `noConstraintType` as an example.

File: src/compiler/checker.ts (hunk 4 above)
Comment (andy-ms) [inline src/compiler/checker.ts]: :+1:

File: src/compiler/checker.ts (hunk 3 above)
Comment (andy-ms) [inline src/compiler/checker.ts]: I want to ensure that this is definitely *not* useable as a `TypePredicate` -- it's just one to prevent polymorphism. Otherwise it's too easy to access `resolvedTypePredicate` and think you're getting the right thing.

File: src/compiler/checker.ts (hunk 5 above)
Comment (andy-ms) [inline src/compiler/checker.ts]: Shouldn't `undefined` be used for the most common case though? Usually no type predicate exists.

Comment (rbuckton) [inline src/compiler/checker.ts]: It can also mean that we don't know the answer yet. This is consistent with other "resolve"-named properties used in checker.

File: src/compiler/checker.ts (hunk 3 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: We generally don't have that issue with other cases like `noConstraintType`, especially if all access to the type predicate is gated through `getTypePredicateOfSignature`.

While I don't think it's strictly necessary, you could add another TypePredicateKind as a discriminant, but I think using an object whose shape (and hidden class) matches other valid values for that property will reduce polymorphism.

File: src/compiler/checker.ts (hunk 5 above)
Comment (andy-ms) [inline src/compiler/checker.ts]: I've inverted the representation, so `undefined` now represents a lazily-computed type predicate and `noTypePredicate` represents no type predicate existing.

File: src/compiler/checker.ts (hunk 3 above)
Comment (andy-ms) [inline src/compiler/checker.ts]: I would disagree that it's not an issue -- this was a barrier to my understanding this code, and it would have been easier had I realized that `resolvedReturnType` was not meant to be used directly as a `Type`. I hadn't come across `noConstraintType` yet, but I'm sure it wouldn't have been easy to realize that a value of type `Type` should not actually be used as a `Type` because it might be a special sentinel value.

Comment (rbuckton) [inline src/compiler/checker.ts]: This one of many cases where we need better documentation in checker.

**Comment by rbuckton** [comment]
Can you take a look at the build failures?

File: src/compiler/checker.ts (hunk 3 above)
Comment (rbuckton) [inline src/compiler/checker.ts]: I would prefer we stick to our current pattern for these cases. We can discuss these specific concerns with the broader team following the 2.5 release.

**Comment by andy-ms** [comment]
@rbuckton @ahejlsberg Any more comments?

**Comment by andy-ms** [comment]
Bump -- @rbuckton @ahejlsberg Is anything blocking this?

**Comment by andy-ms** [comment]
@rbuckton Good to go?

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -6543,6 +6546,20 @@ namespace ts {
...
+
+        function getTypePredicateOfSignature(signature: Signature): TypePredicate | undefined {
+            if (signature.resolvedTypePredicate === noTypePredicate) {
+                return undefined;
+            }
+            if (signature.resolvedTypePredicate === undefined) {
+                signature.resolvedTypePredicate = instantiateTypePredicate(getTypePredicateOfSignature(signature.target), signature.mapper);
+            }
+            return signature.resolvedTypePredicate as TypePredicate;
+        }
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Write this function in the same style as `getConstraintFromTypeParameter`.

**Comment by andy-ms** [comment]
New commit contains fixes to the union half of #17757.
Previously we considered signatures identical even if they had different type predicates; this meant a union of two different type predicates would just contain the first one.

**Comment by andy-ms** [comment]
@sandersn @rbuckton @ahejlsberg Anyone?

File: tests/cases/compiler/typePredicatesInUnion_noMatch.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    pred(x: {}, y: {}): x is boolean;
+}
+interface B {
+    pred(x: {}, y: {}): y is string;
+}
+
+type Or = A | B;
+
+function f(o: Or, x: {}, y: {}) {
+    if (o.pred(x, y)) {
```
[the remark this answers is not in the document]
Comment (sandersn) [reply tests/cases/compiler/typePredicatesInUnion_noMatch.ts]: oh, it's because you can't make a single type predicate that narrows two variables.

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -7503,6 +7563,42 @@ namespace ts {
             return getUnionTypeFromSortedList(typeSet, aliasSymbol, aliasTypeArguments);
         }
 
+        function getUnionTypePredicate(signatures: ReadonlyArray<Signature>): TypePredicate {
+            let first: TypePredicate | undefined;
+            const types: Type[] = [];
+            for (let i = 0; i < signatures.length; i++) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: why not for-of here?

File: tests/cases/compiler/typeInferenceTypePredicate.ts [hunk 9]
```diff
@@ -0,0 +1,3 @@
+declare function f<T>(predicate: (x: {}) => x is T): T;
+// 'res' should be of type 'number'.
```
[the remark this answers is not in the document]
Comment (andy-ms) [reply tests/cases/compiler/typeInferenceTypePredicate.ts]: I didn't want us taking any contextual type -- presumably we should be doing that on assignments if we're not already.

File: src/compiler/checker.ts (hunk 8 above)
Comment (andy-ms) [reply src/compiler/checker.ts]: If we did #19309 we could enable `prefer-for-of` to catch this kind of error.

File: src/compiler/checker.ts [hunk 10]
```diff
@@ -207,7 +207,18 @@ namespace ts {
...
+                getTypePredicateOfSignature,
+                getReturnTypeOfSignature,
+                getBaseTypes,
+                resolveStructuredTypeMembers,
+                getTypeOfSymbol,
+                getResolvedSymbol,
+                getIndexTypeOfStructuredType,
+                getConstraintFromTypeParameter,
+                getFirstIdentifier,
+            ),
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Did anything change here? Otherwise, just leave it alone.

File: src/compiler/checker.ts [hunk 11]
```diff
@@ -228,7 +239,18 @@ namespace ts {
...
+                typeParameters,
+                thisParameter,
+                parameters,
+                resolvedReturnType,
+                resolvedTypePredicate,
+                minArgumentCount,
+                hasRestParameter,
+                hasLiteralTypes) {
+                return createSignature(declaration, typeParameters, thisParameter, parameters, resolvedReturnType, resolvedTypePredicate || noTypePredicate, minArgumentCount, hasRestParameter, hasLiteralTypes);
+            },
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Why the change here?

File: src/compiler/checker.ts [hunk 12]
```diff
@@ -5503,31 +5528,40 @@ namespace ts {
...
+            hasLiteralTypes: boolean,
+        ): Signature {
             const sig = new Signature(checker);
             sig.declaration = declaration;
             sig.typeParameters = typeParameters;
             sig.parameters = parameters;
             sig.thisParameter = thisParameter;
             sig.resolvedReturnType = resolvedReturnType;
-            sig.typePredicate = typePredicate;
+            sig.resolvedTypePredicate = resolvedTypePredicate;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: I would just do `typePredicate || noTypePredicate` here and then you won't need all the other changes.

**Comment by ahejlsberg** [comment]
I just pushed a branch with some suggested changes:

**Comment by andy-ms** [comment]
@ahejlsberg Good to go?

**Comment by ahejlsberg** [comment]
@andy-ms Also, check to see if this fixes #19676.

**Comment by andy-ms** [comment]
#19676 is fixed, tested for by `typeInferenceTypePredicate.ts`. (And tested with local build.)

File: tests/baselines/reference/fixSignatureCaching.types [hunk 13]
```diff
@@ -866,9 +866,9 @@ define(function () {
 >'UnknownMobile' : "UnknownMobile"
 
     isArray = ('isArray' in Array) ?
->isArray = ('isArray' in Array) ?        Array.isArray : function (value) { return Object.prototype.toString.call(value) === '[object Array]'; } : (arg: any) => arg is any[]
+>isArray = ('isArray' in Array) ?        Array.isArray : function (value) { return Object.prototype.toString.call(value) === '[object Array]'; } : (value: any) => boolean
```
Comment (andy-ms) [inline tests/baselines/reference/fixSignatureCaching.types]: I think this change is correct because only the left hand side of the conditional is a type predicate. @ahejlsberg could you verify, then merge

Comment (ahejlsberg) [reply tests/baselines/reference/fixSignatureCaching.types]: Something seems fishy here. If you reverse the order of the operands you get the old result. So apparently both types are subtypes of each other, or otherwise subtype reduction would consistently pick one of them.  I'm not sure why this is the case, since `(x: any) => x is any[]` should be a subtype of `(x: any) => boolean`. You may want to investigate.

Comment (andy-ms) [reply tests/baselines/reference/fixSignatureCaching.types]: Added a test -- the return type seems to be `boolean` the other way around as well.

Comment (ahejlsberg) [reply tests/baselines/reference/fixSignatureCaching.types]: Ok, looks good.

**Comment by ahejlsberg** [comment]
I think this PR is good to go.
